Use Doctrine Skeleton Mapper for the object mapper layer.

jwage opened the pull request:
Playing around with using https://github.com/doctrine/skeleton-mapper for the object mapper layer. This makes it so we can easily manage models where the data is provided from arrays of data provided by container parameters.

SenseException commented:
A lot of files have changed. I'll tests the branch itself too.

SenseException commented on lib/DataSources/TeamMembers.php:
```diff
@@ -0,0 +1,35 @@
+    {
+        $teamMembers = [];
+
+        foreach ($this->teamMembers as $key => $teamMember) {
```
The key is not used here. I kind of expect that the static code analysis would find something like that.

SenseException commented on lib/Model/SitemapPage.php:
```diff
@@ -0,0 +1,50 @@
+    public function hydrate(array $sitemapPage, ObjectManagerInterface $objectManager) : void
+    {
+        $this->url  = $sitemapPage['url'] ?? '';
+        $this->date = $sitemapPage['date'] ?? '';
```
`DateTimeImmutable`. Don't put a string into the date property.

SenseException commented on lib/Model/BlogPost.php:
```diff
@@ -42,6 +46,24 @@ public function __construct(
+    /**
+     * @param mixed[] $project
+     */
+    public function hydrate(array $project, ObjectManagerInterface $objectManager) : void
```
Types for the properties should be handled here or we can get an invalid object state. Too bad `hydrate()` doesn't expect an Exception.

SenseException replied:
I'm looking forward to typed properties too. The hydrator is currently the part where no wrong types should be able to reach its destined property.

SenseException commented on lib/Model/Contributor.php:
```diff
@@ -0,0 +1,91 @@
+    /**
+     * @param mixed[] $contributor
+     */
+    public function hydrate(array $contributor, ObjectManagerInterface $objectManager) : void
```
Types for the properties should be handled in this method or we can get an invalid object state.

SenseException commented on lib/Model/DoctrineUser.php:
```diff
@@ -0,0 +1,43 @@
+    /**
+     * @param mixed[] $doctrineUser
+     */
+    public function hydrate(array $doctrineUser, ObjectManagerInterface $objectManager) : void
```
Types for the properties should be handled in this method or we can get an invalid object state.

SenseException commented on lib/Model/ProjectContributor.php:
```diff
@@ -0,0 +1,97 @@
+    /**
+     * @param mixed[] $projectContributor
+     */
+    public function hydrate(array $projectContributor, ObjectManagerInterface $objectManager) : void
```
Types for the properties should be handled in this method or we can get an invalid object state.

SenseException commented on lib/Model/SitemapPage.php:
```diff
@@ -0,0 +1,50 @@
+    /**
+     * @param mixed[] $sitemapPage
+     */
+    public function hydrate(array $sitemapPage, ObjectManagerInterface $objectManager) : void
```
Types for the properties should be handled in this method or we can get an invalid object state.

SenseException commented on lib/Model/TeamMember.php:
```diff
@@ -0,0 +1,89 @@
+    /**
+     * @param mixed[] $teamMember
+     */
+    public function hydrate(array $teamMember, ObjectManagerInterface $objectManager) : void
```
Types for the properties should be handled in this method or we can get an invalid object state.

SenseException commented on tests/DataSources/BlogPostsTest.php:
```diff
@@ -0,0 +1,95 @@
+
+        $blogPostsData = $this->blogPosts->getData();
+
+        usort($blogPostsData, new Sorter(['date' => 'desc']));
```
Please bring the expected result into the expected order instead of using/introducing a test-unrelated logic.

SenseException replied:
The tested method shouldn't be sorting at all and the mock object controls in what order the result is returned, so the test shouldn't have a surprise order. A DataSource doesn't sort any values, so I think that even `array_reverse` isn't needed in `BlogPosts`. The sorting will be done by the Repository.

jwage commented on composer.json:
```diff
@@ -13,9 +13,13 @@
+        "doctrine/skeleton-mapper": "dev-master",
         "erusev/parsedown": "^1.7",
+        "knplabs/github-api": "^2.10",
+        "php-http/guzzle6-adapter": "^1.1",
```
It was required by the github client.

Unrelated but this is why the team data was empty for you. You need to configure the github api key. I have it catching the exception right now. I need to remove that and fix it so I can configure the api key on travis so the tests don't fail.

jwage commented on config/services.xml:
```diff
@@ -33,10 +33,27 @@
         <service id="Symfony\Component\Console\Application" autowire="true" />
         <service id="Symfony\Component\Filesystem\Filesystem" autowire="true" />
+        <service id="Doctrine\Common\Cache\FilesystemCache" autowire="false">
+            <argument>%doctrine.website.cache_dir%</argument>
```
I think I didn't use auto wiring here because it is not a class I control so the argument names are not very specific.

jwage replied:
The FilesystemCache constructor looks like:

    public function __construct($directory, $extension = self::EXTENSION, $umask = 0002)

So I would have to bind a pretty ambiguous argument variable name. It felt better to be explicit here. Thoughts?

jwage commented on lib/DataSource/DataSourceObjectDataRepository.php:
```diff
@@ -0,0 +1,142 @@
+use function in_array;
+use function usort;
+
+class DataSourceObjectDataRepository extends BasicObjectDataRepository
```
In the skeleton mapper you have two interfaces. The first is how you provide the raw data to the mapper called ObjectDataRepository. The other is ObjectRepository which you are familiar with from the other Doctrine object mapper implementations.

jwage replied:
This is maybe a signal the naming could be better, but that is how it is today in doctrine/skeleton mapper.

jwage commented on lib/DataSource/Sorter.php:
```diff
@@ -0,0 +1,86 @@
+use function sprintf;
+use function strtolower;
+
+class Sorter
```
I will add tests for everything in this PR before we merge.

jwage commented:
I renamed `DataSource::getData()` to `DataSource::getSourceRows()` but I am not sure about it either. It is better, but I am still not 100% happy with it.

The tests fail now because I removed the catching of the exception when the GitHub API fails. I need to figure out how to use TravisCI environment variables so that I can set the github API key so it can retrieve all the team/contributors data properly.

SenseException commented on lib/DataSources/ProjectContributors.php:
```diff
@@ -0,0 +1,112 @@
+
+                $teamMember = $this->teamMemberRepository->findOneByGithub($contributor['author']['login']);
+
+                $isMaintainer = $teamMember !== null
```
This line uses the new variable too.

jwage replied:
phpstan doesn't like it. I guess it isn't smart enough to handle this case right now.

```
1032 /data/doctrine-website (object-mapper) $ phpstan a
Note: Using configuration file /data/doctrine-website/phpstan.neon.dist.
 122/122 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%

 ------ -------------------------------------------------------------------------------------
  Line   lib/DataSources/ProjectContributors.php
 ------ -------------------------------------------------------------------------------------
  68     Cannot call method isProjectMaintainer() on Doctrine\Website\Model\TeamMember|null.
 ------ -------------------------------------------------------------------------------------


 [ERROR] Found 1 error
```

SenseException replied:
Yeah, I tried some combinations but it looks like phpstan doesn't understand this case.

SenseException commented:
I'll take a look at it soon. :+1:
